### PR TITLE
Inlocomedia adapter

### PR DESCRIPTION
**Pull request by anshumansworld**
This includes support of native, interstitial & feed banner ads for InLocoMedia network

### CODE REVIEW

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkFeedBannerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+        Log.v(TAG, "destroy");
+        if (mFeedBanner != null) {
+            mFeedBanner.destroy();
+        }
+    }
+
+    @Override
+    public void hide() {
+        if (mFeedBanner.getParent() != null) {
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkFeedBannerAdapter.java]: missing log

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkInterstitialAdapter.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    public InLocoMediaNetworkInterstitialAdapter(Map data) {
+
+        super(data);
+    }
+
+    @Override
+    public void load(Context context) {
+
+        Log.v(TAG, "load");
+        if (context != null && mData != null) {
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkInterstitialAdapter.java]: I prefer positive comparisons, It's less tricky to read
```
if (context == null || mData == null) {
  invokeLoadFail(PubnativeException.ADAPTER_ILLEGAL_ARGUMENTS);
} else {
 // REQUEST
}
```

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkFeedBannerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    //==============================================================================================
+    // Public
+    //==============================================================================================
+
+    @Override
+    public void load(Context context) {
+
+        Log.v(TAG, "load");
+        if (context != null && mData != null) {
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkFeedBannerAdapter.java]: LEt's use possitive comparison instead

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    //==============================================================================================
+    // PubnativeNetworkAdapter methods
+    //==============================================================================================
+
+    @Override
+    protected void request(Context context) {
+
+        Log.v(TAG, "request");
+        if (context != null && mData != null) {
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: positive comparison instead

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            String appId = (String) mData.get(KEY_APP_ID);
+            String adUnitId = (String) mData.get(KEY_AD_UNIT_ID);
+            if (TextUtils.isEmpty(appId) || TextUtils.isEmpty(adUnitId)) {
+                invokeFailed(PubnativeException.ADAPTER_ILLEGAL_ARGUMENTS);
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: thhis should be missing data

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            } else {
+                createRequest(context, appId, adUnitId);
+            }
+        } else {
+            invokeFailed(PubnativeException.ADAPTER_MISSING_DATA);
```
Comment (cerberillo) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: this should be illegal arguments

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import android.view.View;
+import android.view.ViewGroup;
+
+import com.google.android.gms.ads.formats.NativeAppInstallAd;
+import com.google.android.gms.ads.formats.NativeAppInstallAdView;
+import com.inlocomedia.android.ads.AdView;
+
+import net.pubnative.mediation.request.model.PubnativeAdModel;
+
+public class InLocoMediaNativeAdModel extends PubnativeAdModel {
```
Comment (cerberillo) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java]: This empty class here does nothing, why?

If InLocoMedia doesn't allows native ads, then the hub should return null on the native adapter

File: mediation/src/main/java/net/pubnative/mediation/adapter/network/PubnativeNetworkInterstitialAdapter.java [hunk 8]
```diff
@@ -169,7 +169,6 @@ protected void invokeLoadFinish() {
         if (mLoadListener != null) {
             mLoadListener.onAdapterLoadFinish(this);
         }
-        mLoadListener = null;
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/adapter/network/PubnativeNetworkInterstitialAdapter.java]: load listener should be nullified once the load operation is completed

File: mediation/src/main/java/net/pubnative/mediation/request/model/PubnativeAdModel.java [hunk 9]
```diff
@@ -140,6 +140,15 @@ public void setListener(PubnativeAdModel.Listener listener) {
...
     public abstract View getAdvertisingDisclosureView(Context context);
 
+    /**
+     * gets the ad view
+     *
+     * @return Ad view.
+     */
+    public ViewGroup getAdView() {
+        return null;
+    }
```
Comment (cerberillo) [inline mediation/src/main/java/net/pubnative/mediation/request/model/PubnativeAdModel.java]: what is this for?

File: mediation.demo/src/main/java/net/pubnative/mediation/demo/AdViewHolder.java [hunk 10]
```diff
@@ -121,23 +126,30 @@ public void renderAd() {
...
-            Picasso.with(mContext).load(model.getIconUrl()).into(mIcon);
-            Picasso.with(mContext).load(model.getBannerUrl()).into(mBanner);
-            View sponsorView = model.getAdvertisingDisclosureView(mContext);
-            if (sponsorView != null) {
-                mAdDisclosure.addView(sponsorView);
+            mCustomAdView.removeAllViews();
+            mCustomAdView.setVisibility(View.GONE);
+            if(model.getAdView() != null) {
+                mCustomAdView.addView(model.getAdView(), 0, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
+                mCustomAdView.setVisibility(View.VISIBLE);
```
Comment (cerberillo) [inline mediation.demo/src/main/java/net/pubnative/mediation/demo/AdViewHolder.java]: why is this needed? can't we stick to the old model?

File: mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    // InLocoMediaNetworkAdapter methods
+    //==============================================================================================
+    protected void createRequest(Context context, String appId, String adUnitId) {
+
+        Log.v(TAG, "createRequest");
+        InLocoMediaOptions options = InLocoMediaOptions.getInstance(context);
+        options.setAdsKey(appId);
+        InLocoMedia.init(context, options);
+
+        AdRequest adRequest = new AdRequest();
```
Comment (RagnarR) [inline mediation.adaters/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: How about targeting for `AdRequest`?

File: mediation.demo/src/main/res/layout/ad_list_cell.xml [hunk 12]
```diff
@@ -49,6 +49,13 @@
             android:id="@+id/ad_view_container"
             android:layout_below="@+id/ad_adapter_name_text">
 
+            <RelativeLayout
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:id="@+id/customAdView"
```
Comment (RagnarR) [inline mediation.demo/src/main/res/layout/ad_list_cell.xml]: Please, put ID as first line. That's Google recommendation and it easier for reading.

File: mediation/src/main/java/net/pubnative/mediation/adapter/network/PubnativeNetworkInterstitialAdapter.java (hunk 8 above)
Comment (anshumansworld) [inline mediation/src/main/java/net/pubnative/mediation/adapter/network/PubnativeNetworkInterstitialAdapter.java]: If so how other methods like impression confirmed, clicked would be invoked?

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java (hunk 7 above)
Comment (anshumansworld) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java]: It allows native ads but the thing is it defines it's own ids for each element & take a layout that contains all the required views(ids) in it. So the implementation is like:
We inflate our layout while making request and set this custom layout in `mAdView` that could be retrieved at the time of displaying ads in the client app. App will have to handle this by checking if `mAdView` is not null, add this ad in a container and tracking would be handled within InLocoMedia SDK.

We need to update this kind of usage in readme for InLocoMedia network.

File: mediation.demo/src/main/java/net/pubnative/mediation/demo/AdViewHolder.java (hunk 10 above)
Comment (RagnarR) [inline mediation.demo/src/main/java/net/pubnative/mediation/demo/AdViewHolder.java]: I think the better way to use something like `withCustomAdView` for `model` and then put all data inside. Like we implement for AdMob. How do you think?

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        adRequest.setAdUnitId(adUnitId);
+        return adRequest;
+    }
+
+    //==============================================================================================
+    // CALLBACKS
+    //==============================================================================================
+    // AdViewListener
+    //----------------------------------------------------------------------------------------------
+    protected class NativeAdListener extends AdViewListener {
```
Comment (RagnarR) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: I prefer create variable from the extended class instead creating an inner class. By the way, inner class not very good for performance. You can use something like:
```
protected AdViewListener listener = new AdViewListener() {
@Override
         public void onAdViewReady(AdView adView) {
 
             Log.v(TAG, "onAdViewReady");
             InLocoMediaNativeAdModel wrapper = new InLocoMediaNativeAdModel(adView);
             invokeLoaded(wrapper);
         }
 
         @Override
         public void onAdError(AdView adView, com.inlocomedia.android.ads.AdError adError) {
 
             Log.v(TAG, "onAdError");
             Map errorData = new HashMap();
             errorData.put("adError", adError.toString());
             invokeFailed(PubnativeException.extraException(PubnativeException.ADAPTER_UNKNOWN_ERROR, errorData));
         }
};
```

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java (hunk 7 above)
Comment (RagnarR) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java]: I not sure, that it right way for putting the view inside a constructor. You could use `startTracking()` method for setting your view, I think. Because we using constructor for setting native data in other models.

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java (hunk 13 above)
Comment (cerberillo) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: Really? I don't like this approach, using the inner class as a representation instead of a variable makes it more present in the code, in fact that's the way we're using across our products

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java (hunk 7 above)
Comment (cerberillo) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/model/InLocoMediaNativeAdModel.java]: @anshumansworld I think this is a pretty special integration and we need it for tomorrow as much, so I'll take care of this task now on

File: mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java (hunk 13 above)
Comment (RagnarR) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: @cerberillo inner class not very well for leaking memory. [Here](http://blog.nimbledroid.com/2016/05/23/memory-leaks.html) is a good description of it.

Comment (cerberillo) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: @RagnarR can you point the specific place where it's described why is it a leak?

Comment (RagnarR) [inline mediation.demo/src/main/java/net/pubnative/mediation/adapter/network/InLocoMediaNetworkRequestAdapter.java]: @cerberillo item 3 in article:
> What if we create an instance of this Inner Class and maintain a static reference to it? At this point you might as well just guess that a memory leak is imminent. Unfortunately because one of the benefits of Inner Class instances is that they have access to their Outer Class’s variables, they must maintain a reference to the Outer Class’s instance which causes our Activity to be leaked.

**Comment by cerberillo** [comment]
This PR is stopped since we don't think that inloco SDK can be successfully mediated using our current structure, they are using a very unique setup and generally speaking they don't provide native ads, but configurable in feed ads, we shoujld be able to get the inner data and be able to capture impression and clicks events

**Comment by cerberillo** [comment]
I'm closing this PR because we need to solve some blockers before merging this andwe will probably require some work in other branches before going again with this
